react/week2/to_do_list/src/index.js

```diff
@@ -0,0 +1,177 @@
+import React from "react";
+import ReactDOM from "react-dom";
+import "./index.css";
+
+const taskList = [
+  {
+    id: 0,
+    description: "Get out of bed",
+    color: "brown",
+    deadline: "2020-06-08",
+    done: false,
+    visibility: true
+  },
+  {
+    id: 1,
+    description: "Brush teeth",
+    color: "red",
+    deadline: "2020-06-08",
+    done: false,
+    visibility: true
+  },
+  {
+    id: 2,
+    description: "Eat breakfast",
+    color: "orange",
+    deadline: "2020-06-08",
+    done: true,
+    visibility: true
+  }
+];
+
+class Header extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1> Welcome to My ToDo List</h1>
+        <br ></br>
+      </div>
+    )
+  }
+}
+
+class Counter extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      ellapsedTime: this.props.initial,
+    };
+  }
+
+  componentDidMount() {
+    setInterval(this.increment, 1000);
+  }
+
+  increment = () => {
+    this.setState({ ellapsedTime: this.state.ellapsedTime + 1 })
+  }
+
+  StartTimer = () => {
+    setInterval(this.increment, 1000);
+  }
+
+  render() {
+
+    return (
+      <div>
+        <p> you have spent {this.state.ellapsedTime} seconds in this page</p>
+        <br></br>
+      </div>
+    )
+  }
+}
+
+class TodoList extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = {
+      toDoListData: this.props.toDoListData
+    }
+  }
+
+  addRandomTask(toDoListData) {
+    toDoListData.push({
+      id: toDoListData.length,
+      description: "Random Text",
+      color: 'navygrey',
+      deadline: "2020-12-29",
+      done: false,
+      visibility: true
+    });
+    this.setState({ toDoListData: this.props.toDoListData })
+  }
+
+  render() {
+    let toDoListJSXarray = [];
+    for (let i = 0; i < this.props.taskList.length; i++) {
+      toDoListJSXarray.push(<ToDoTask task={this.props.taskList[i]} />)
+    }
+    return (
+      <div>
+        <button onClick={e => this.addRandomTask(this.props.taskList)}> Add a Task </button>
```

Comment:
I don't think this `addRandomTask` operates as you'd expect. It looks like you pass in props, mutate props, and then set the state as the current props anyways, effectively not changing anything. 

It's very important to remember props are **read-only** meaning they can't be updated. Props should be used to pass data to components, state should be used for maintaining dynamic data that needs to be updated: https://reactjs.org/docs/components-and-props.html#props-are-read-only

It's certainly a nuanced concept, so I would recommend watching as many videos as you can on the topic to better understand it!

Reply:
Oh, You are so sharp.... I got your point. Thank you. I fixed it.